### PR TITLE
mimir: merge of ways having same name and admins with level = city_level

### PR DESCRIPTION
**Pull request by kadhikari**
Need https://github.com/CanalTP/mimirsbrunn/pull/65 to be merged to add
some tests.

Ticket concerned : http://jira.canaltp.fr/browse/NAVP-590

### CODE REVIEW

File: src/bin/osm2mimir.rs [hunk 1]
```diff
@@ -106,6 +113,11 @@ impl AdminMatcher {
     }
 }
 
+#[derive(Debug, Clone, Ord, PartialOrd, Eq, PartialEq)]
+pub struct StreetKey {
+	pub name: String,
+	pub admins: BTreeSet<Rc<mimir::Admin>>,
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: You can use `AdminSet` here.

File: src/bin/osm2mimir.rs [hunk 2]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+    let mut street_rel : StreetWithRelationSet = BTreeSet::new();
+    let mut street_list: StreetsVec = vec![];
     // Sometimes, streets can be devided into several "way"s that still have the same street name.
     // The reason why a street is devided may be that a part of the street become a bridge/tunne/etc.
     // In this case, a "relation" tagged with (type = associatedStreet) is used to group all these "way"s.
-    // In order not to have doublons in autocompleion, we should keep only one "way" in the relation
-    // and remove all the rest way whose street name is the same.
-    let mut objs_to_remove = Vec::<osmpbfreader::OsmId>::new();
-
+    // In order not to have duplicates in autocompleion, we should keep only one "way" in the relation
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: In order not to have duplicates in autocompletion, we should tag the osm ways use in the relation to not index them twice.

File: src/bin/osm2mimir.rs [hunk 3]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+    
     for (_, rel_obj) in &objs_map {
         if let &osmpbfreader::OsmObj::Relation(ref rel) = rel_obj {
-            let mut found = false;
+            //TODO : create a street for the first memeber way of the relation but not for the rest.
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: This TODO is done.

File: src/bin/osm2mimir.rs [hunk 4]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+                    let admin = get_street_admin(admins_geofinder, objs_map, way);
+                    ret ret(street_list.push(mimir::Street {
+							id: way.id.to_string(),
+							street_name: way_name.to_string(),
+                    		label: way_name.to_string(),
+                    		weight: 1,
+                    		administrative_regions: admin,
+            			}))
                 };
+                break;
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: you can `}.unwrap_or(break);` to break only if we inserted a street.

File: src/bin/osm2mimir.rs [hunk 5]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+   	    		let admins_geofinder = &admins_geofinder;
+                
+                mdo! {
+                    when ref_obj.member.is_way();
+                    obj =<< objs_map.get(&ref_obj.member);
+                    let way = match obj {
+                        &osmpbfreader::OsmObj::Way(ref w) => w,
+                        _ => unreachable!()
+                    };
+                    way_name =<< way.tags.get("name");
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: maybe the relation name is better if it exists?

File: src/bin/osm2mimir.rs (hunk 4 above)
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: Nop, doesn't work, but maybe
```rust
}.unwrap_or(continue);
break;
```

File: src/bin/osm2mimir.rs [hunk 6]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+							street_name: way_name.to_string(),
+                    		label: way_name.to_string(),
+                    		weight: 1,
+                    		administrative_regions: admin,
+            			}))
                 };
+                break;
+            }
+            
+            //Add osmid of all the relation members in de set street_rel.
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: in _the_set

File: src/bin/osm2mimir.rs [hunk 7]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+            //We don't create any street for all the osmid present in street_rel  
+            for ref_obj in &rel.refs {
+                if let &osmpbfreader::OsmId::Way(_) = &ref_obj.member {
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]:
```rust
if ref_obj.member.is_way() {
```

File: src/bin/osm2mimir.rs [hunk 8]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+                    street_rel.insert(ref_obj.member.clone());
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: `OsmId` is `Copy`, thus you can just
```rust
street_rel.insert(ref_obj.member);
```

File: src/bin/osm2mimir.rs [hunk 9]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+    
+    // we merge all the ways with a key = way_name + admin list of level(=city_level)
+    //we use a map NameAdminMap <key, value> to manage the merging of ways
+    let mut name_admin_map : NameAdminMap = BTreeMap::new();
+    for (osmid, obj) in &objs_map {
+        if street_rel.contains(osmid) {
+        	continue;    
+        }
+        if let &osmpbfreader::OsmObj::Way(ref way) = obj {
+            let admin: BTreeSet<Rc<mimir::Admin>> = get_street_admin(&admins_geofinder, &objs_map, way).into_iter().filter(|admin| admin.level == city_level).collect();
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: long line

File: src/bin/osm2mimir.rs (hunk 1 above)
Comment (kadhikari) [inline src/bin/osm2mimir.rs]: ok

File: src/bin/osm2mimir.rs (hunk 2 above)
Comment (kadhikari) [inline src/bin/osm2mimir.rs]: ok

File: src/bin/osm2mimir.rs (hunk 3 above)
Comment (kadhikari) [inline src/bin/osm2mimir.rs]: ok

File: src/bin/osm2mimir.rs (hunk 5 above)
Comment (kadhikari) [inline src/bin/osm2mimir.rs]: ok

File: src/bin/osm2mimir.rs [hunk 10]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+             if let Some(way_name) = way.tags.get("name") {
+                 name_admin_map.entry(StreetKey{name: way_name.to_string(), admins: admin}).or_insert(vec![]).push(*osmid)
+             } 
+        }
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]:
```rust
use mdo::option::*;
mdo! {
    way =<< obj.way();
    let admins: BtreeSet<_> = get_street_admin(&admins_geofinder, &objs_map, way)
        .into_iter()
        .filter(|admin| admin.level == city_level)
        .collect();
    way_name =<< way.tags.get("name");
    let key = StreetKey { name: way_name.to_string(), admins: admins };
    ret ret(name_admin_map.entry(key).or_insert(vec![]).push(*osmid))
}
```

File: src/bin/osm2mimir.rs [hunk 11]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+   	//Create a street for each way with osmid present in in objs_map 
+   	for (_, way_ids) in name_admin_map {
+   	    use mdo::option::*;
+   	    let objs_map = &objs_map;
+   	    let street_list = &mut street_list;
+   	    let admins_geofinder = &admins_geofinder;
+   	    mdo! {
+			obj =<< objs_map.get(&way_ids[0]);
+			way =<< obj.way();
+   	        //way =<< way(obj);
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: strange indent and useless comment.

File: libs/mimir/src/objects.rs.in [hunk 12]
```diff
@@ -102,6 +102,26 @@ pub struct Admin {
...
+    }
+}
+
+impl PartialEq for Admin {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+impl Eq for Admin {}
```
Comment (xlqian) [inline libs/mimir/src/objects.rs.in]: you still need this trait?

File: src/bin/osm2mimir.rs (hunk 5 above)
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]:
```rust
way_name =<< rel.tag.get("name").or_else(|| way.tags.get("name"));
```

File: src/bin/osm2mimir.rs [hunk 13]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
-                    }
-                    objs_to_remove.push(ref_obj.member.clone());
+                
+                use mdo::option::*;
+   	    		let objs_map = &objs_map;
+   	    		let street_list = &mut street_list;
+   	    		let admins_geofinder = &admins_geofinder;
+                
+                mdo! {
+                    when ref_obj.member.is_way();
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]:
```rust
_ =<< if ref_obj.role == "street" { Some(()) } else { None };
```

Comment (TeXitoi) [inline src/bin/osm2mimir.rs]:
```rust
when ref_obj.role == "street"
```
it should work.

File: src/bin/osm2mimir.rs [hunk 14]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
-                    if !found {
-                        found = true;
-                        continue;
-                    }
-                    objs_to_remove.push(ref_obj.member.clone());
+                
+                use mdo::option::*;
+   	    		let objs_map = &objs_map;
+   	    		let street_list = &mut street_list;
+   	    		let admins_geofinder = &admins_geofinder;
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: bad indentation

File: src/bin/osm2mimir.rs [hunk 15]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+    //we use a map NameAdminMap <key, value> to manage the merging of ways
+    let mut name_admin_map : NameAdminMap = BTreeMap::new();
+    for (osmid, obj) in &objs_map {
+        if street_rel.contains(osmid) {
+        	continue;    
+        }
+        if let &osmpbfreader::OsmObj::Way(ref way) = obj {
+            let admin: BTreeSet<Rc<mimir::Admin>> = get_street_admin(&admins_geofinder, &objs_map, way).into_iter().filter(|admin| admin.level == city_level).collect();
+             if let Some(way_name) = way.tags.get("name") {
+                 name_admin_map.entry(StreetKey{name: way_name.to_string(), admins: admin}).or_insert(vec![]).push(*osmid)
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: long line

File: src/bin/osm2mimir.rs [hunk 16]
```diff
@@ -240,55 +253,95 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+			way =<< obj.way();
+   	        //way =<< way(obj);
+   	        way_name =<< way.tags.get("name");
+   	        let admins = get_street_admin(admins_geofinder, objs_map, way);
+   	        ret ret(street_list.push(mimir::Street{
+					id: way.id.to_string(),
+					street_name: way_name.to_string(),
+                	label: way_name.to_string(),
+                	weight: 1,
+                	administrative_regions: admins,
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: bad indent

**Comment by TeXitoi** [comment]
Great work!

File: libs/mimir/src/objects.rs.in [hunk 17]
```diff
@@ -102,6 +102,26 @@ pub struct Admin {
...
+}
+
+impl PartialOrd for Admin {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for Admin {
+    fn eq(&self, other: &Self) -> bool {
```
Comment (TeXitoi) [inline libs/mimir/src/objects.rs.in]: https://doc.rust-lang.org/std/cmp/trait.PartialEq.html

File: src/bin/osm2mimir.rs (hunk 6 above)
Comment (kadhikari) [inline src/bin/osm2mimir.rs]: ok

File: src/bin/osm2mimir.rs (hunk 4 above)
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: OK, then 
```rust
let inserted = mdo! {
   ...
};
if inserted.is_some() { break; }
```

File: src/bin/osm2mimir.rs (hunk 7 above)
Comment (kadhikari) [inline src/bin/osm2mimir.rs]: ok

File: src/bin/osm2mimir.rs (hunk 8 above)
Comment (kadhikari) [inline src/bin/osm2mimir.rs]: ok

File: src/bin/osm2mimir.rs (hunk 9 above)
Comment (kadhikari) [inline src/bin/osm2mimir.rs]: ok

File: libs/mimir/src/objects.rs.in (hunk 17 above)
Comment (xlqian) [inline libs/mimir/src/objects.rs.in]: ok

File: libs/mimir/src/objects.rs.in (hunk 12 above)
Comment (TeXitoi) [inline libs/mimir/src/objects.rs.in]: `Ord` want `Eq`, and `Ord` is needed for `BTreeSet`

**Comment by TeXitoi** [comment]
You need a rebase.

File: src/bin/osm2mimir.rs (hunk 11 above)
Comment (kadhikari) [inline src/bin/osm2mimir.rs]: ok

File: src/bin/osm2mimir.rs [hunk 18]
```diff
@@ -236,55 +247,105 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
             _ => false,
         }
     };
-
     info!("reading pbf...");
-    let mut objs_map = osmpbfreader::get_objs_and_deps(pbf, is_valid_obj).unwrap();
-    info!("reading pbf done.");
+    let objs_map = osmpbfreader::get_objs_and_deps(pbf, is_valid_obj).unwrap();
+	info!("reading pbf done.");
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: tab indent.

File: src/bin/osm2mimir.rs [hunk 19]
```diff
@@ -236,55 +247,105 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+    let mut street_rel : StreetWithRelationSet = BTreeSet::new();
+    let mut street_list: StreetsVec = vec![];    
     // Sometimes, streets can be devided into several "way"s that still have the same street name.
     // The reason why a street is devided may be that a part of the street become a bridge/tunne/etc.
     // In this case, a "relation" tagged with (type = associatedStreet) is used to group all these "way"s.
-    // In order not to have doublons in autocompleion, we should keep only one "way" in the relation
-    // and remove all the rest way whose street name is the same.
-    let mut objs_to_remove = Vec::<osmpbfreader::OsmId>::new();
-
+    // In order not to have duplicates in autocompleion, we should tag the osm ways in the relation not to index them twice.
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: long line.

File: src/bin/osm2mimir.rs [hunk 20]
```diff
@@ -236,55 +247,105 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+   	    		let admins_geofinder = &admins_geofinder;
+   	    		
+                let inserted = mdo! {
+                    when ref_obj.member.is_way();
+                    when ref_obj.role == "street";
+                    obj =<< objs_map.get(&ref_obj.member);
+                    let way = match obj {
+                        &osmpbfreader::OsmObj::Way(ref w) => w,
+                        _ => unreachable!()
+                    };
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]:
```rust
way =<< obj.way();
```

File: src/bin/osm2mimir.rs [hunk 21]
```diff
@@ -236,55 +247,105 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+                    way_name =<< way_name.or_else(|| way.tags.get("name"));
+                    let admin = get_street_admin(admins_geofinder, objs_map, way);
+                    ret ret(street_list.push(mimir::Street {
+							id: way.id.to_string(),
+							street_name: way_name.to_string(),
+                    		label: way_name.to_string(),
+                    		weight: 1,
+                    		administrative_regions: admin,
+            			}))
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: bad indentation

File: src/bin/osm2mimir.rs [hunk 22]
```diff
@@ -236,55 +247,105 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+   	//Create a street for each way with osmid present in in objs_map 
+   	for (_, way_ids) in name_admin_map {
+		use mdo::option::*;
+		let objs_map = &objs_map;
+		let street_list = &mut street_list;
+		let admins_geofinder = &admins_geofinder;
+   	    mdo! {
+			obj =<< objs_map.get(&way_ids[0]);
+			way =<< obj.way();
+   	        way_name =<< way.tags.get("name");
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: Bad indent

File: src/bin/osm2mimir.rs [hunk 23]
```diff
@@ -236,55 +247,105 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+   	        let admins = get_street_admin(admins_geofinder, objs_map, way);
+   	        ret ret(street_list.push(mimir::Street{
+				id: way.id.to_string(),
+				street_name: way_name.to_string(),
+				label: way_name.to_string(),
+                weight: 1,
+                administrative_regions: admins,
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: One more time

File: src/bin/osm2mimir.rs [hunk 24]
```diff
@@ -303,10 +364,8 @@ fn main() {
     info!("creating adminstrative regions");
     let admins = administrative_regions(&mut parsed_pbf, levels);
 
-    info!("creating streets");
-    let mut streets = streets(&mut parsed_pbf, &admins);
+    let mut streets = streets(&mut parsed_pbf, &admins, city_level);
     
-    info!("computing city weight");
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: You've removed a lot of messages.

File: src/bin/osm2mimir.rs [hunk 25]
```diff
@@ -236,55 +249,103 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec) -> StreetsVec {
...
+            way_name =<< way.tags.get("name");
+            let admins = get_street_admin(admins_geofinder, objs_map, way);
+            ret ret(street_list.push(mimir::Street{
+   	                    id: way.id.to_string(),
+   	                    street_name: way_name.to_string(),
+   	                    label: way_name.to_string(),
+   	                    weight: 1,
+   	                    administrative_regions: admins,
+            }))
+   	    };
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: here

**Comment by TeXitoi** [comment]
OK for me.

**Pull request merged**